easybuild/framework/easyblock.py

```diff
@@ -3039,41 +3039,48 @@ def run_all_steps(self, run_test_cases):
         print_msg("building and installing %s..." % self.full_mod_name, log=self.log, silent=self.silent)
         trace_msg("installation prefix: %s" % self.installdir)
 
-        lockpath = build_option('lockpath') or os.path.join(install_path('software'), '.locks')
-        if not os.path.exists(lockpath):
-            mkdir(lockpath, parents=True)
-        lockfile_name = os.path.join(lockpath, ".%s.lock" % self.installdir.replace('/', '_'))
-        if os.path.exists(lockfile_name):
-            if build_option('wait_on_lock'):
-                while os.path.exists(lockfile_name):
-                    print_msg("Lock file %s exists. Waiting 60 seconds." % lockfile_name, silent=self.silent)
-                    time.sleep(60)
-            else:
-                print_msg("Build aborted. Lock file %s exists." % lockfile_name, silent=self.silent)
-                return False
+        ignore_locks = build_option('ignore_locks')
+
+        if ignore_locks:
+            self.log.info("Ignoring locks...")
         else:
-            try:
-                # create a new lock file
-                print_msg("Creating lock file %s" % lockfile_name, silent=self.silent)
-                f = open(lockfile_name, "w+")
-                f.close()
-
-                for (step_name, descr, step_methods, skippable) in steps:
-                    if self._skip_step(step_name, skippable):
-                        print_msg("%s [skipped]" % descr, log=self.log, silent=self.silent)
+            locks_dir = build_option('locks_dir') or os.path.join(install_path('software'), '.locks')
+            lock_path = os.path.join(locks_dir, '%s.lock' % self.installdir.replace('/', '_'))
+
+            # if lock already exists, either abort or wait until it disappears
+            if os.path.exists(lock_path):
+                wait_on_lock = build_option('wait_on_lock')
+                if wait_on_lock:
+                    while os.path.exists(lock_path):
+                        print_msg("lock %s exists, waiting %d seconds..." % (lock_path, wait_on_lock),
+                                  silent=self.silent)
+                        time.sleep(wait_on_lock)
+                else:
+                    raise EasyBuildError("Lock %s already exists, aborting!", lock_path)
+
+            # create lock to avoid that another installation running in parallel messes things up;
+            # we use a directory as a lock, since that's atomically created
+            mkdir(lock_path, parents=True)
```

Comment:
We probably need to catch a failure of creating the directory

Reply:
The `mkdir` function has a `try`/`except`, and will print a proper error when the lock dir could not be created:

```
$ eb bzip2-1.0.6.eb -f --locks-dir /
== temporary log file in case of crash /tmp/eb-jlbj8B/easybuild-N6ikuA.log
== processing EasyBuild easyconfig /Users/kehoste/work/easybuild-easyconfigs/easybuild/easyconfigs/b/bzip2/bzip2-1.0.6.eb
== building and installing bzip2/1.0.6...
== FAILED: Installation ended unsuccessfully (build directory: /tmp/kehoste/bzip2/1.0.6/system-system): build failed (first 300 chars): Failed to create directory /_Users_kehoste_software_bzip2_1.0.6.lock: [Errno 30] Read-only file system: '/_Users_kehoste_software_bzip2_1.0.6.lock' (took 0 sec)
== Results of the build can be found in the log file(s) /tmp/eb-jlbj8B/easybuild-bzip2-1.0.6-20200330.162332.bfdHw.log
ERROR: Build of /Users/kehoste/work/easybuild-easyconfigs/easybuild/easyconfigs/b/bzip2/bzip2-1.0.6.eb failed (err: "build failed (first 300 chars): Failed to create directory /_Users_kehoste_software_bzip2_1.0.6.lock: [Errno 30] Read-only file system: '/_Users_kehoste_software_bzip2_1.0.6.lock'")
```

Reply:
I think that error should be caught in the code, and issue a clean error message and then exit from the code, rather than escalating outside of the function.